diff --git a/VrApp.gradle.org b/VrApp.gradle
index 5195ab1..835f12d 100755
--- a/VrApp.gradle.org
+++ b/VrApp.gradle
@@ -95,7 +95,7 @@ class VrAppPlugin implements Plugin<Project> {
     // FIXME: The Task.leftShift(Closure) method has been deprecated and is scheduled to be removed in Gradle 5.0. Please use Task.doLast(Action) instead.
     project.task( "cleanWorkAround" ) {
 	  description "Workaround for .externalNativeBuild not being deleted on clean"
-    } << {
+    } doLast {
 	  project.delete project.file( ".externalNativeBuild" )
     }
 
@@ -198,7 +198,7 @@ class VrAppPlugin implements Plugin<Project> {
 
   project.afterEvaluate {
 
-      Task initDeviceList = project.task( "initDeviceList()" ) << {
+      Task initDeviceList = project.task( "initDeviceList()" ) doLast {
         project.ext.deviceMap = [ : ]
         if (project.hasProperty( "should_install" ) == true || project.hasProperty( "disable_sig_check" ) == false) {
             AndroidDebugBridge.initIfNeeded( false )
@@ -225,7 +225,7 @@ class VrAppPlugin implements Plugin<Project> {
 
       project.task( "stopApk", dependsOn: initDeviceList ) {
           description "Stops app if currently running on device"
-      } << {
+      } doLast {
           project.deviceMap.each { deviceSerial, device ->
           stopApk( device, android.defaultConfig.applicationId )
         }
@@ -237,7 +237,7 @@ class VrAppPlugin implements Plugin<Project> {
             project.deviceMap.size() != 0 &&
             project.hasProperty( "disable_sig_check" ) == false
           }
-      } << {
+      } doLast {
           def iterator = project.deviceMap.entrySet().iterator()
           while ( iterator.hasNext() ) { 
             def deviceElem = iterator.next()
@@ -264,7 +264,7 @@ class VrAppPlugin implements Plugin<Project> {
             project.hasProperty( "disable_sig_check" ) == false
           }
           dependsOn project.OSigPreCheck
-        } << { variant.outputs.each { output ->
+        } doLast { variant.outputs.each { output ->
             def sigfiles = project.zipTree( output.outputFile ).matching{ include "assets/oculussig*" }
             def iterator = project.deviceMap.entrySet().iterator()
             while ( iterator.hasNext() ) { 
@@ -285,7 +285,7 @@ class VrAppPlugin implements Plugin<Project> {
           dependsOn OSigPostCheck
           onlyIf { project.hasProperty( "should_install" ) }
           description "Installs and runs the APK file"
-        } << { variant.outputs.each { output ->
+        } doLast { variant.outputs.each { output ->
             if ( output.outputFile.exists() )
             {
               if ( project.deviceMap.size() == 0 ) {
@@ -317,7 +317,7 @@ class GenDebugKeystore extends DefaultTask {
     @TaskAction
     genDebugKeystore() {
         if ( !project.file("android.debug.keystore").exists() ) {
-            def pythonCmd = 'python'
+            def pythonCmd = 'python2'
             if (Os.isFamily(Os.FAMILY_WINDOWS)) {
                 pythonCmd = "${project.rootProject.projectDir}/bin/scripts/build/ovrbuild_keystore.py.bat"
                 project.exec {
@@ -356,7 +356,7 @@ class BuildSDKLibs extends DefaultTask {
                 buildType
             }
         } else {
-            def pythonCmd = 'python'
+            def pythonCmd = 'python2'
             project.exec {
                 commandLine pythonCmd,
                 "${project.rootProject.projectDir}/bin/scripts/build/ovrbuild_sdklibs.py",
@@ -396,4 +396,4 @@ project.tasks.whenTaskAdded { task ->
     }
 }
 
-apply plugin: VrAppPlugin
\ No newline at end of file
+apply plugin: VrAppPlugin
diff --git a/VrLib.gradle.org b/VrLib.gradle
index 343968e..b996f87 100755
--- a/VrLib.gradle.org
+++ b/VrLib.gradle
@@ -6,7 +6,7 @@ class VrLibPlugin implements Plugin<Project> {
 
     project.task( "cleanWorkAround" ) {
       description "Workaround for .externalNativeBuild not being deleted on clean"
-    } << {
+    } doLast {
       project.delete project.file( ".externalNativeBuild" )
     }
 
